samza-core/src/main/java/org/apache/samza/startpoint/StartpointManager.java

```diff
@@ -0,0 +1,273 @@
+ * 1) Only by {@link SystemStreamPartition}
+ * 2) A combination of {@link SystemStreamPartition} and {@link TaskName}
+ *
+ * The intention for the StartpointManager is to maintain a strong contract between the caller
+ * and how Startpoints are stored in the underlying MetadataStore.
+ */
+@InterfaceStability.Unstable
+public class StartpointManager {
+  private static final Logger LOG = LoggerFactory.getLogger(StartpointManager.class);
+  private static final String NAMESPACE = "samza-startpoint-v1";
+
+  private final StartpointSerde startpointSerde = new StartpointSerde();
+
+  private MetadataStore metadataStore;
+
+  private StartpointManager(MetadataStore metadataStore) {
+    this.metadataStore = metadataStore;
+  }
+
+  /**
+   * Creates a {@link StartpointManager} instance.
+   * @param config {@link Config} required for the underlying store.
+   * @param metricsRegistry {@link MetricsRegistry} to hook into the underlying store.
+   * @return {@link StartpointManager} instance.
+   */
+  public static StartpointManager getInstance(Config config, MetricsRegistry metricsRegistry) {
+    Preconditions.checkNotNull(config, "Config cannot be null");
+    Preconditions.checkNotNull(metricsRegistry, "MetricsRegistry cannot be null");
+
+    String metadataStoreFactoryName = new JobConfig(config).getStartpointMetadataStoreFactory();
+    MetadataStore metadataStore;
+    try {
+      MetadataStoreFactory metadataStoreFactory = Util.getObj(metadataStoreFactoryName, MetadataStoreFactory.class);
+      metadataStore = metadataStoreFactory.getMetadataStore(NAMESPACE, config, metricsRegistry);
+      metadataStore.init();
+    } catch (ConfigException | NullPointerException ex) {
+      LOG.error("StartpointManager cannot be created because the underlying metadata store could not be configured or created.", ex);
+      return null;
+    }
+
+    StartpointManager startpointManager = new StartpointManager(metadataStore);
+    LOG.info("StartpointManager created with metadata store: {}",
+        startpointManager.metadataStore.getClass().getCanonicalName());
+
+    return startpointManager;
+  }
+
+  /**
+   * Writes a {@link Startpoint} that defines the start position for a {@link SystemStreamPartition}.
+   * @param ssp The {@link SystemStreamPartition} to map the {@link Startpoint} against.
+   * @param startpoint Reference to a Startpoint object.
+   */
+  public void writeStartpoint(SystemStreamPartition ssp, Startpoint startpoint) {
+    Preconditions.checkNotNull(metadataStore, "Underlying metadata store not available");
+    Preconditions.checkNotNull(ssp, "SystemStreamPartition cannot be null");
+    Preconditions.checkNotNull(startpoint, "Startpoint cannot be null");
+
+    try {
+      metadataStore.put(toStoreKey(ssp),
+          startpointSerde.toBytes(startpoint.copyWithStoredAt(Instant.now().toEpochMilli())));
+    } catch (Exception ex) {
+      throw new SamzaException(String.format(
```

Comment:
Why do we need to catch the exception here wrap the exception and throw it back. What is the value-add we get compared to  throwing up the base exception back  to the caller?